tests/api.cpp

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file
+ * @brief This file provides unit tests for the ArgoDSM API functions
+ * @copyright Eta Scale AB. Licensed under the Eta Scale Open Source License. See the LICENSE file for details.
+ */
+
+#include <iostream>
+
+#include <limits.h>
+#include <unistd.h>
+
+#include "argo.hpp"
+#include "allocators/collective_allocator.hpp"
+#include "backend/backend.hpp"
+#include "env/env.hpp"
+#include "data_distribution/data_distribution.hpp"
+#include "gtest/gtest.h"
+
+/** @brief ArgoDSM memory size */
+constexpr std::size_t size = 1<<26;
+/** @brief ArgoDSM cache size */
+constexpr std::size_t cache_size = size/8;
+
+namespace env = argo::env;
+namespace dd = argo::data_distribution;
+namespace mem = argo::mempools;
+extern mem::global_memory_pool<>* default_global_mempool;
+
+/** @brief ArgoDSM page size */
+const std::size_t page_size = 4096;
+
+/** @brief A "random" char constant */
+constexpr char c_const = 'a';
+
+/**
+ * @brief Class for the gtests fixture tests. Will reset the allocators to a clean state for every test
+ */
+class APITest : public testing::Test {
+
+	protected:
+		APITest() {
+			argo_reset();
+			argo::barrier();
+		}
+
+		~APITest() {
+			argo::barrier();
+		}
+};
+
+
+
+/**
+ * @brief Unittest that checks correctness of the
+ * argo::is_argo_address(T* addr) API function.
+ */
+TEST_F(APITest, IsArgoAddress) {
+	std::size_t alloc_size = default_global_mempool->available();
+	char *tmp = static_cast<char*>(collective_alloc(alloc_size));
+	char *global_base = argo::backend::global_base();
+	std::size_t global_size = argo::backend::global_size();
+	ASSERT_FALSE(argo::is_argo_address(global_base-1));
+	ASSERT_TRUE(argo::is_argo_address(global_base));
+	ASSERT_TRUE(argo::is_argo_address(&tmp[0]));
+	ASSERT_TRUE(argo::is_argo_address(&tmp[alloc_size/2]));
+	ASSERT_TRUE(argo::is_argo_address(&tmp[alloc_size-1]));
+	ASSERT_FALSE(argo::is_argo_address(global_base+global_size));
+}
+
+/**
+ * @brief Unittest that checks correctness of the
+ * argo::get_homenode(T* addr) API function.
+ */
+TEST_F(APITest, GetHomeNode) {
+	std::size_t alloc_size = default_global_mempool->available();
+	char *tmp = static_cast<char*>(collective_alloc(alloc_size));
+	argo::node_id_t node_id = argo::node_id();
+	std::size_t num_nodes = argo::number_of_nodes();
+	char* start = argo::backend::global_base();
+	char* end = start + argo::backend::global_size();
+
+	/* Touch an equal number of pages per node */
+	int first_writer = (argo::get_homenode(start)+1) % num_nodes;
+	int my_start = (node_id-first_writer+num_nodes)%num_nodes;
+	for(std::size_t s=page_size*my_start;
+			s<alloc_size-1;
+			s+=page_size*num_nodes) {
+		tmp[s] = c_const;
+	}
+	argo::barrier();
```

Comment:
does this imply that prefetching must not cause a homenode to be set if the prefetched page is still unused?

Reply:
Prefetching causes a homenode to be set, whether this is correct behavior or not can be discussed. This is just to ensure that each page is owned by a specific node regardless of prefetching (since it can be disabled through envvars).

Reply:
Actually, this is not true, mistake on my end! Pages that do not have a homenode are not prefetched at all unless they are the target of the cache miss.

Reply:
then I do not understand how this allows to "Test that the number of pages owned by each node is equal" below. wouldn't prefetching change how many pages are owned per node?

Reply:
Probably a bad explanation on my end.

Loading a page `p` that does not have an owner claims ownership of `p` on behalf of the loading node, since we know that we want to access this page in some way.
However, loading of the subsequent `ARGO_LOAD_SIZE-1` pages is subject to whether the pages are
1. owned by the owner of `p` (same homenode)
2. contiguous in backing store (`offset of p+1 == offset of p + page_size`)

If one of these does not hold true, prefetching of this and subsequent pages stops. A page that does not have an owner fails the first condition and prefetching stops.

Reply:
>2) is a condition that only ever fails in the first-touch policy.

if I fetch the final page in one of node0's cyclic allocation blocks, does it prefetch into node1, or does it fetch a block aligned to the end of node0's block, or does it fail the condition and stop prefetching?

Reply:
Under the current design, it fails condition 1 (homenode) on the following page and stops prefetching.